src/plugins/data/public/data_sources/datasource_services/datasource_service.ts

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { BehaviorSubject } from 'rxjs';
+import {
+  IDataSourceFilters,
+  IDataSourceRegisterationResult,
+  DataSourceRegisterationError,
+  GenericDataSource,
+} from './types';
+
+export class DataSourceService {
+  private static dataSourceService: DataSourceService;
+  // A record to store all registered data sources, using the data source name as the key.
+  private dataSources: Record<string, GenericDataSource> = {};
+  private _dataSourcesSubject: BehaviorSubject<Record<string, GenericDataSource>>;
+
+  private constructor() {
+    this._dataSourcesSubject = new BehaviorSubject(this.dataSources);
+  }
+
+  static getInstance(): DataSourceService {
+    if (!this.dataSourceService) {
+      this.dataSourceService = new DataSourceService();
+    }
+    return this.dataSourceService;
+  }
```

Comment:
Not the best way to do singletons.

Reply:
discussing this in https://github.com/opensearch-project/OpenSearch-Dashboards/pull/5167#discussion_r1346955915